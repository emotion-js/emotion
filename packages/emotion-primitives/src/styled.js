// @flow
import * as React from 'react'
import { View, Text, Image, StyleSheet } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign
} from './utils'
import { getStyles } from './getStyles'
import { convertToRNStyles } from './convertToRNStyles'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

// Evaluate the styles and convert them to React Native using styled component context
function evalStyles(context, Comp, styles) {
  // Convert styles with or without interpolations to React Native (StyleSheet.create)
  let otherStyles = convertToRNStyles.call(context, styles)

  return getStyles.call(context, otherStyles, context.props)
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
export function createStyled(component: React.ElementType) {
  let isPrimitive = isPrimitiveComponent(component)
  let pick = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...styles: *) {
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })

        const emotionStyles = evalStyles(this, Styled, styles)

        return React.createElement(
          component,
          pickAssign(pick, {}, props, {
            ref: props.innerRef,
            style: StyleSheet.flatten(emotionStyles)
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
