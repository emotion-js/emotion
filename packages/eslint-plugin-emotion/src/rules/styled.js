module.exports = {
  meta: {
    fixable: 'code'
  },
  create(context) {
    return {
      ImportDeclaration(node) {
        if (
          node.source.value === 'react-emotion' ||
          node.source.value === 'preact-emotion'
        ) {
          let newImportPath =
            node.source.value === 'react-emotion'
              ? '@emotion/styled'
              : '@emotion/preact-styled'
          let quote = node.source.raw.charAt(0)
          context.report({
            node: node.source,
            message: `styled should be imported from ${newImportPath}`,
            fix:
              node.specifiers.length === 1 &&
              node.specifiers[0].type === 'ImportDefaultSpecifier'
                ? fixer => {
                    return fixer.replaceText(
                      node.source,
                      quote + newImportPath + quote
                    )
                  }
                : undefined
          })
        }
      }
    }
  }
}
