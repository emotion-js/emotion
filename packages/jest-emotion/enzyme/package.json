{
  "main": "dist/jest-emotion.cjs.js",
  "module": "dist/jest-emotion.esm.js",
  "types": "../types/enzyme",
  "preconstruct": {
    "source": "../src/enzyme"
  }
}
