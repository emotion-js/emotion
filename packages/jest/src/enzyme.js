// @flow
export { createEnzymeSerializer } from './create-enzyme-serializer'
export { matchers } from './matchers'
