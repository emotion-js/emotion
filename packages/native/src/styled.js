import { StyleSheet } from 'react-native'
import { createStyled } from '@emotion/primitives-core'

/**
 * a function that returns a styled component which render styles in React Native
 */
let styled = createStyled(StyleSheet)

export { styled }
