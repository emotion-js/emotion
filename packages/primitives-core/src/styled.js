// @flow
import * as React from 'react'
import { testAlwaysTrue, pickAssign, interleave } from './utils'
import { ThemeContext } from '@emotion/core'
import { createCss } from './css'

let defaultPickTest = prop => prop !== 'theme' && prop !== 'innerRef'

type options = {
  getShouldForwardProp: (cmp: React.ElementType) => (prop: string) => boolean
}

export function createStyled(
  StyleSheet: Object,
  { getShouldForwardProp = () => defaultPickTest }: options = {}
) {
  const css = createCss(StyleSheet)

  return function createEmotion(component: React.ElementType) {
    let pickTest = getShouldForwardProp(component)

    return function createStyledComponent(...rawStyles: *) {
      let styles

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles)
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      // $FlowFixMe
      let Styled = React.forwardRef((props, ref) => {
        return (
          <ThemeContext.Consumer>
            {theme => {
              let mergedProps = pickAssign(testAlwaysTrue, {}, props, {
                theme: props.theme || theme
              })
              let stylesWithStyleProp = styles
              if (props.style) {
                stylesWithStyleProp = styles.concat(props.style)
              }
              const emotionStyles = css.apply(mergedProps, stylesWithStyleProp)

              if (process.env.NODE_ENV !== 'production' && props.innerRef) {
                console.error(
                  'innerRef is no longer supported, please use ref instead'
                )
              }

              return React.createElement(
                component,
                pickAssign(pickTest, {}, props, {
                  ref: ref,
                  style: emotionStyles
                })
              )
            }}
          </ThemeContext.Consumer>
        )
      })
      // $FlowFixMe
      Styled.withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      Styled.displayName = `emotion(${getDisplayName(component)})`

      return Styled
    }
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
