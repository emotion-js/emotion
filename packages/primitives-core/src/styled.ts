import * as React from 'react'
import { interleave } from './utils'
import { ThemeContext } from '@emotion/react'
import { createCss } from './css'
import { AbstractStyleSheet } from './types'

let testOmitPropsOnComponent = (prop: string) =>
  prop !== 'theme' && prop !== 'as'

interface CreateStyledOptions {
  getShouldForwardProp(cmp: React.ElementType): (prop: string) => boolean
}

interface StyledOptions {
  shouldForwardProp?(prop: string): boolean
}

type StyledProps = Record<string, unknown> & {
  as?: React.ElementType
}

export function createStyled(
  StyleSheet: AbstractStyleSheet,
  options?: CreateStyledOptions
) {
  const getShouldForwardProp =
    options?.getShouldForwardProp ?? (() => testOmitPropsOnComponent)

  const css = createCss(StyleSheet)

  return function createEmotion(
    component: React.ElementType,
    options?: StyledOptions
  ) {
    let shouldForwardProp =
      options && options.shouldForwardProp
        ? options.shouldForwardProp
        : undefined
    let defaultShouldForwardProp =
      shouldForwardProp || getShouldForwardProp(component)
    let shouldUseAs = !defaultShouldForwardProp('as')

    return function createStyledComponent(...rawStyles: any[]) {
      let styles: any[]

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles as [any, ...any[]])
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      let Styled = React.forwardRef<unknown, StyledProps>((props, ref) => {
        const finalTag = (shouldUseAs && props.as) || component

        let mergedProps = props
        if (props.theme == null) {
          mergedProps = {}
          for (let key in props) {
            mergedProps[key] = props[key]
          }
          mergedProps.theme = React.useContext(ThemeContext)
        }

        let finalShouldForwardProp =
          shouldUseAs && shouldForwardProp === undefined
            ? getShouldForwardProp(finalTag)
            : defaultShouldForwardProp

        let newProps: Record<string, unknown> = {}

        for (let key in props) {
          if (shouldUseAs && key === 'as') continue

          if (finalShouldForwardProp(key)) {
            newProps[key] = props[key]
          }
        }

        newProps.style = [css.apply(mergedProps, styles), props.style]
        newProps.ref = ref

        return React.createElement(finalTag, newProps)
      })

      Styled.displayName = `emotion(${getDisplayName(component)})`

      const withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      const castedStyled = Styled as typeof Styled & {
        withComponent: typeof withComponent
      }

      castedStyled.withComponent = withComponent

      return castedStyled
    }
  }
}

const getDisplayName = (
  primitive: string | { displayName?: string; name?: string }
) =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
