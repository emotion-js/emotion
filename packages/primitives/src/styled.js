// @flow
import * as React from 'react'
import { View, Text, Image } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign,
  interleave
} from './utils'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'
import { css } from './css'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */

type CreateStyledComponent = (...styles: any) => React.ElementType

type BaseCreateStyled = (tag: React.ElementType) => CreateStyledComponent

export type CreateStyled = BaseCreateStyled & {
  View: CreateStyledComponent,
  Text: CreateStyledComponent,
  Image: CreateStyledComponent
}

// $FlowFixMe
let createStyled: CreateStyled = function createStyled(
  component: React.ElementType
) {
  let isPrimitive = isPrimitiveComponent(component)
  let pickTest = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...rawStyles: *) {
    let styles

    if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
      styles = rawStyles
    } else {
      styles = interleave(rawStyles)
    }
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })
        let stylesWithStyleProp = styles
        if (props.style) {
          stylesWithStyleProp = styles.concat(props.style)
        }
        const emotionStyles = css.apply(this, stylesWithStyleProp)

        return React.createElement(
          component,
          pickAssign(pickTest, {}, props, {
            ref: props.innerRef,
            style: emotionStyles
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'

export { createStyled }
