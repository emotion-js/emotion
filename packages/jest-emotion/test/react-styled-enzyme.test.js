import 'test-utils/legacy-env' // adds serializer
/** @jsx jsx */
import * as enzyme from 'enzyme'
import { jsx } from '@emotion/core'
import styled from '@emotion/styled'
import toJson from 'enzyme-to-json'

function toMatchSnapshot(component) {
  const wrapper = enzyme.mount(component)
  expect(toJson(wrapper)).toMatchSnapshot()
}

beforeEach(() => {
  document.getElementsByTagName('html')[0].innerHTML = ''
})

it('Styled test 1', () => {
  const Component = styled.div``

  toMatchSnapshot(<Component />)
})

// this seems to kick some sort of initialization
// TODO: investigate
it('Styled test 2', () => {
  const Component = styled.div`
    color: red;
  `

  toMatchSnapshot(<Component />)
})

it('Styled test 3', () => {
  const Component = styled.div``

  toMatchSnapshot(<Component />)
})
