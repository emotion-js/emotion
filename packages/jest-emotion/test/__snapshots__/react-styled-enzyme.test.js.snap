// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled test 1 1`] = `
<Component>
  <div
    className="css-1mndwhu-Component emotion-0"
  />
</Component>
`;

exports[`Styled test 2 1`] = `
<Component>
  <div
    className="css-1scxj8d-Component emotion-0"
  />
</Component>
`;

exports[`Styled test 3 1`] = `
<Component>
  <div
    className="css-1mndwhu-Component emotion-0"
  />
</Component>
`;
